src/chainstate/stacks/index/trie.rs

```diff
@@ -64,42 +75,77 @@ fn get_nodetype_hash<T: MarfTrieId>(
 
     let ret = TrieHash(res);
 
-    trace!("get_node_hash: hash {:?} = {:?} + ::children::", &ret, node);
+    trace!(
+        "get_nodetype_hash: hash {:?} = {:?} + ::children::",
+        &ret,
+        node
+    );
     Ok(ret)
 }
 
 impl Trie {
     /// Read the root node.  First try to read it as a back-pointer (since all root nodes except for
     /// the root node in the very first trie will be back-pointers), and if that fails due to a
     /// node ID mismatch (i.e. CorruptionError), then try to read it as a non-backpointer.
-    pub fn read_root<T: MarfTrieId>(
+    #[inline]
+    fn read_root_maybe_hash<T: MarfTrieId>(
         storage: &mut TrieStorageConnection<T>,
+        read_hash: bool,
     ) -> Result<(TrieNodeType, TrieHash), Error> {
```

Comment:
Using `Option<TrieHash>` and unwrapping from the caller is a more explicit (and slightly safer) way to represent what's happening here than using empty buffers. It would also avoid the allocations (though that's admittedly probably a pretty small cost).

Reply:
Sure, can do